client/src/components/RecipeForm.js

```diff
@@ -0,0 +1,198 @@
+import React, { useState } from "react";
+
+const RecipeForm = (props) => {
+  
+  const [recipePayload, setRecipePayload] = useState({
+    name: '',
+    ingredients: [{name: ''}],
+    instructions: [{name: ''}],
+    notes: '',
+    tags: [],
+    image: '',
+
+  })
+
+  const [errors, setErrors] = useState({});
+
+  const [shouldRedirect, setShouldRedirect] = useState(false);
+
+  const onInputChange = (event) => {
+    setRecipePayload({
+      ...recipePayload,
+      [event.currentTarget.name]: event.currentTarget.value
+    })
+  }
+
+  const onIngredientChange = index => event => {
+    const newIngredients = recipePayload.ingredients.map((ingredient, ingredientIndex) => {
+      if (index != ingredientIndex) return ingredient
+      return {...ingredient, name: event.target.value}
+    })
+    setRecipePayload({
+      ...recipePayload,
+      ingredients: newIngredients
+    })
+  }
+
+  const onInstructionChange = index => event => {
+    const newInstructions = recipePayload.instructions.map((instruction, instructionIndex) => {
+      if (index != instructionIndex) return instruction
+      return {...instruction, name: event.target.value}
+    })
+    setRecipePayload({
+      ...recipePayload,
+      instructions: newInstructions
+    })
+  }
+
+  const addIngredient = () => {
+    setRecipePayload({
+      ...recipePayload,
+      ingredients: recipePayload.ingredients.concat([{ name: '' }])
+    })
+  }
+
+  const removeIngredient = (index) => () => {
+    setRecipePayload({
+      ...recipePayload,
+      ingredients: recipePayload.ingredients.filter((ingredient, ingredientIndex) => index != ingredientIndex)
+    })
+  }
+
+  const addInstruction = () => {
+    setRecipePayload({
+      ...recipePayload,
```

Comment:
How does the `...recipePayload` syntax work?  Unfamiliar with typescript.

Reply:
The spread `...` syntax lets us include the current state of recipePayload when setting its state, so that we don't lose its current state when making changes.

For this function, when we addInstruction(), we set the state of recipePayload to its current state PLUS the new element in recipePayload.instructions.

From MDN: Spread syntax can be used when all elements from an object or array need to be included in a new array or object

Reply:
nice, thanks!